src/translator/service.cpp

```diff
@@ -202,6 +313,11 @@ void AsyncService::pivot(std::shared_ptr<TranslationModel> first, std::shared_pt
 void AsyncService::translate(std::shared_ptr<TranslationModel> translationModel, std::string &&source,
                              CallbackType callback, const ResponseOptions &responseOptions) {
   // Producer thread, a call to this function adds new work items. If batches are available, notifies workers waiting.
+  // Tagging
+  for (auto&& teminologyPair : terminologyMap_) {
+    source = Replace(source, teminologyPair.first, teminologyPair.second);
+  }
```

Comment:
Wait so `terminologyForce` is exactly the same as doing it on the input yourself? I.e.

```py
def forceTerminology(text, terms):
  for term, replacement in sorted(terms.items(), reverse=True, key=lambda pair: len(pair[0])):
    text = text.replace(term, replacement)
  return text


translator = Translator(...)
translator.translate([forceTerminology(text, terminology)])
 ```

That also highlights two issues I have with your implementation:

1. I don't see any explicit sorting on the length of the to be replaced terms. In that case, if you have `house` and `houseboat` in there, and the first one occurs first in the map, `houseboat` is never matched because `house` will already have been replaced.
2. Because you're restarting every loop from the start, terminology that appears in the replacement part of one term can then be modified if it matches the search part of another term.